### ISSUE TITLE
Re-design: Improve error handling and reporting in ingest processes

### ISSUE THREAD

**Issue by nmacgreg** [ualbertalib/discovery#1875]
**Resulting from a problem**
See: https://helpdesk.library.ualberta.ca/otrs/index.pl?Action=AgentTicketZoom;TicketID=55157
When the cron job for ingesting updates to SFX / eJournals data failed last night, it highlighted some limitations with this code and the environment surrounding it: 

1. The output of the cron/rake job is difficult to process _mechanically_
2. The job removes the existing data, before analysing whether the download of new, replacement was successful
While we have a monitoring solution (Nagios queries the data, to see if any is missing) to detect this situation, it should be possible to leave the existing data in place, while alerting the appropriate staff that the daily update process is failing. 

**Towards a solution**
1. The rake tasks for ingest need comprehensive, mature error handling. For instance, if the task fails to download the new data file, it should stop, report the problem, and not remove the old data, nor try to ingest a file that doesn't exist.
2. Whenever errors occur, clearly and consistently report errors, in a form that can be read both mechanically (eg by Nagios) and that makes it clear to operators how to resolve the problem.
3. When designing an improved solution, learn about the features of our monitoring system (or interface with experts), and take them into account, making it easy for Nagios to accurately pinpoint the problem and alert the staff.

**Additional Context**
Whether sfx/databases/symphony data, and whether run from the command line or a cron job, error detection, handling, and reporting should be consistent, and integrated with our monitoring system, so that operator attention can be efficiently directed at problems!

**Comment by pgwillia** [comment]
From @nmacgreg's email "Re: missing journals"

Given the high impact of this problem to our patrons, the recent frequency of troubles, and the long, long time that these problems have existed in the codebase, I hope you set the priority of this issue very high!

Can I question design choices, again? I've brought this up before... there's probably still a GH issue...

- Why does the output of the job contain confusing, blatant error messages even when it works correctly?  This makes it hard to parse both by humans, and automation - eg:
```
INFO [main] (MarcImporter.java:749) -  Connecting to remote Solr server at URL http://solrcloud-prod.library.ualberta.ca:8080/solr/discovery-prod/update
    java.io.IOException: Server returned HTTP response code: 500 for URL: http://solrcloud-prod.library.ualberta.ca:8080/solr/discovery-prod/admin/registry.jsp
```
- Why is the actual error message not better highlighted, in the log output? (Again, making it hard to parse) - eg:
```
INFO [main] (MarcImporter.java:577) - Exception occurred while Indexing: Unable to parse input
```
- Why is the job designed in such a way that it discards perfectly good data (although it's only 1 day old), before knowing that the new data is consistent?


The best I can do, from a Systems perspective, is already done: Nagios is **correctly** monitoring whether the data is present or not (don't throw shade at Nagios!)

This job (and sister-rake-task for [databases]), could use some TLC!  I suggest ideas for improving the design:
- **never** delete data from Solr that is required for normal operation ... Solr supports the concept of overwriting, does it not?  (the monthly rebuild of the SolrCloud index would clear out old data)
  - missing data is bad!  Day-old data (and the knowledge that last night's cron job failed) would be much, much better! 
  - (... but, what's the impact for data that's deleted at the source, but not overwritten/deleted by this cronjob/rake task) or, at least: pre-download and pre-parse the data file to check for data consistency, before deleting existing data, if you find it MUST be deleted
 - or, at least: does SolrCloud support the concept of transactions?  Can you make the delete-and-re-write operation atomic, such that you can roll it back, in the event of trouble late in the operation?
 - or, at least: can you **keep the datafile from the last successful ingest**? and re-ingest it, if the current update fails?  (again, this needs to be combined with a Nagios report that the cron job is failing, to attract Admin attention)
- at the very, very least, we need to have a conversation about better integrating this with monitoring!
  - Ideally, Nagios needs to report the **state of the most recent run of the cronjobs**: success or failure (in addition to testing whether data is missing and causing errors).  This would make it more clear to oncall staff (Henry) that the root cause of the problem is a failed cronjob, and clear up confusion about whether this is a problem with Nagios, a problem with Solr, or a problem with the application code throwing 500's.  But the current state of the output in the log file makes this impossibly hard.  Ideas - choose one:
    - touch a file indicating success, at the end of the job  eg I'll create a new Nagios job that expects the file to exist, & have a recent timestamp
      - (delete this file, at the beginning of the job)
      - (maybe write "pending" into the file at the beginning, so Nagios can monitor whether it's taking too long)(or maybe manipulate a file indicating pending status)
      - (or, touch a file indicating failure; maybe write a one-liner error into it)
    - "fix" the output recorded in the log file: clearly report success or failure, so I can write new code in Nagios to parse the output...  Hints:
      - use a consistent keyword, like "ERROR" or "FAIL", or "SUCCESS"
  - Nagios could potentially re-run a job that fails, and fix the problem before anyone gets beeped out of bed


Last night's log attached...
[sfxIngest.log](https://github.com/ualbertalib/discovery/files/4956013/sfxIngest.log)

**Comment by pgwillia** [comment]
We could keep a backup of the old data before we fetch here
https://github.com/ualbertalib/discovery/blob/edd287abc8a7bf48c376f520406b341efe34091b/lib/tasks/fetch.rake#L4-L8

We could try to parse the xml before trying to run it through SolrMarc 
https://github.com/ualbertalib/discovery/blob/408ca66e6829e832603557885f65dca6ff08173c/lib/tasks/ingest.rake#L48
```
doc = File.open(Rails.root.join('data', 'sfx.xml')) { |f| Nokogiri::XML(f) }
doc.errors
=> [#<Nokogiri::XML::SyntaxError: 428:24: FATAL: Input is not proper UTF-8, indicate encoding !
Bytes: 0xFC 0x73 0x73 0x65>]
```
